### PR TITLE
feat(molot): BREAKING CHANGE Replaced plugin loaders and override mec…

### PR DESCRIPTION
**Pull request by thebruce**
BREAKING CHANGE Replaced plugin loaders and override mechanism with molotov.
This PR replaces the schemePunk plugin loaders with the molotov module's plugin loaders. 
It also revises the testing framework to Jest.
Finally it prepares the codebase for use in node 6 and 8.

### CODE REVIEW

File: lib/supers/destination.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+   * @param {Object} scheme - A destination scheme.
+   * @param {Object} holdOvers - An object of holdovers.
+   */
+  init(options, transformedValue, scheme, holdOvers) {
+    // eslint-disable-next-line no-unused-vars
+    this.scheme = scheme;
+    this.setTarget(options.target);
+    this.setDestination(options.destinationValue);
+    this.setValue(transformedValue);
+    this.setHoldOvers(holdOvers);
```
Comment (elliotttf) [inline lib/supers/destination.js]: If this should be async, you should return a `Promise.resolve(this)` here, allowing for a pattern like this: 

```javascript
new Thing().init().then(inst => inst.doNextThing())
```

File: lib/destination/schemePunkDestination.js [hunk 2]
```diff
@@ -1,154 +1,29 @@
...
-   * validateDestinationTarget - checks to see if the target exists.
-   *
-   * @return boolean
-   */
-  validateDestinationTarget() {
-    if ({}.hasOwnProperty.call(this.scheme[this.getDestination()], this.getTarget())) {
-      return true;
+  return setUp.then(extendedClass => class schemePunkDestination extends extendedClass {
+    init(options, transformedValue, scheme, holdOvers) {
+      super.init(options, transformedValue, scheme, holdOvers);
```
[the remark this answers is not in the document]
Comment (thebruce) [reply lib/destination/schemePunkDestination.js]: agreed!

File: lib/plugins/source/jsonTemplateFileSource.js [hunk 3]
```diff
@@ -24,17 +24,15 @@ module.exports = superclass => class extends superclass {
...
-      trace => trace.getFileName() === base
-    );
+    let traceIndex = stackTrace
+      .findIndex(trace => trace.getFileName() === base);
     traceIndex = this.getTraceIndex(traceIndex);
 
-    this.retrievedOrigin = fs.readJsonSync(
-      path.resolve(
+    this.retrievedOrigin = fs
+      .readJsonSync(path.resolve(
```
[the remark this answers is not in the document]
Comment (thebruce) [reply lib/plugins/source/jsonTemplateFileSource.js]: Set up #93 to handle this. I think I need a rewrite of setters to return this for chaining and consistent handling of async operations.

File: lib/plugins/transform/typeAdapter.js [hunk 4]
```diff
@@ -35,24 +35,24 @@ module.exports = superclass => class extends superclass {
...
+  * |####      ___)    #|
+  * |######           ##|
+  * \##### ---------- /
+  * \####           (
+  * `\###          |
+  * \###         |
+  * \##        |
+  * \###.    .)
+  * `======/
+  */
```
[the remark this answers is not in the document]
Comment (thebruce) [reply lib/plugins/transform/typeAdapter.js]: I like what you got!

File: lib/schemePunk.js [hunk 5]
```diff
@@ -47,50 +47,68 @@ module.exports = class schemePunk {
...
+      .then((SchemeSource) => {
         const holdOvers = Object.assign({}, this.holdOvers);
         // Create a new schemePunk factory.
-        const SchemeSource = schemePunkSourceBase(
-          this.options.source,
-          schemePunkScheme
-        );
-        const schemeSource = new SchemeSource(
+        const schemeSource = new SchemeSource();
+        schemeSource.init(
```
[the remark this answers is not in the document]
Comment (thebruce) [reply lib/schemePunk.js]: absolutely! changed for source, transform and destination.

File: lib/supers/destination.js (hunk 1 above)
Comment (thebruce) [reply lib/supers/destination.js]: done for each of the classes source, transform, and destination.

File: .npmignore [hunk 6]
```diff
@@ -0,0 +1 @@
+__tests__
```
Comment (elliotttf) [inline .npmignore]: So apparently the presence of `.npmignore` overrides `.gitignore`, so if there's anything else (e.g. `__coverage__`) that you don't want to publish to npm you should add it here too.

**Comment by thebruce** [comment]
Version [3.0.0](https://github.com/schemepunk/scheme-punk/releases/tag/3.0.0) has been released.

**Pull request merged**